### PR TITLE
add atp tests

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

I wrote my tests here.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: atp/protocol_test.go
Comment (jaredoconnell) [inline atp/protocol_test.go:1]: I think this could use more comments explaining the reasoning behind each case.

File: atp/client.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ type Client interface {
 	ReadSchema() (schema.Schema[schema.Step], error)
 	// Execute executes a step with a given context and returns the resulting output.
 	Execute(ctx context.Context, stepID string, input any) (outputID string, outputData any, err error)
+	Encoder() *cbor.Encoder
+	Decoder() *cbor.Decoder
```
Comment (jaredoconnell) [inline atp/client.go:29]: Is this necessary outside of testing? It's not always a good idea to expose everything.

File: atp/server.go [hunk 2]
```diff
@@ -91,11 +100,49 @@ func RunATPServer( //nolint:funlen
...
+
+type ATPServer struct {
+	channel ServerChannel
+	decMode cbor.DecMode
+	logger  log.Logger
+	Decoder *cbor.Decoder
+	Encoder *cbor.Encoder
+}
+
+func NewATPServer(
```
Comment (jaredoconnell) [inline atp/server.go:128]: Is this only for the test cases at the moment?

**Review by jaredoconnell — CHANGES_REQUESTED** [review]

File: atp/client.go (hunk 1 above)
Comment (mfleader) [reply atp/client.go:29]: Afaik, it is not necessary outside of testing. It appeared to be the only solution to create deterministic edge cases.

What do you suggest instead?

File: atp/server.go (hunk 2 above)
Comment (mfleader) [reply atp/server.go:128]: Yep

**Review by jaredoconnell — APPROVED** [review]
Looks good. We will need to more thoroughly test this before merging into main.

**Pull request merged**